File: api/src/routes/devices.py
from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from src.db import get_db, now_utc
from src.models import DeviceAssign, DeviceCreate, DeviceResponse

router = APIRouter(prefix="/api/v1/devices", tags=["devices"])


@router.post("", status_code=201, response_model=DeviceResponse)
async def register_device(body: DeviceCreate, request: Request):
    db = get_db(request)
    existing = await db.query_one("SELECT id FROM devices WHERE id = ?", (body.id,))
    if existing:
        return JSONResponse(
            status_code=409,
            content={
                "error": "conflict",
                "message": "Device already registered",
            },
        )
    now = now_utc()
    await db.execute(
        "INSERT INTO devices (id, name, created_at, updated_at) VALUES (?, ?, ?, ?)",
        (body.id, body.name, now, now),
    )
    return await db.query_one("SELECT * FROM devices WHERE id = ?", (body.id,))


@router.get("", response_model=dict)
async def list_devices(request: Request):
    db = get_db(request)
    rows = await db.query("SELECT * FROM devices ORDER BY created_at DESC")
    return {"items": rows}


@router.post("/{device_id}/assign", response_model=DeviceResponse)
async def assign_device(device_id: str, body: DeviceAssign, request: Request):
    db = get_db(request)
    device = await db.query_one("SELECT * FROM devices WHERE id = ?", (device_id,))
    if not device:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "message": "Device not found",
            },
        )
    batch = await db.query_one("SELECT * FROM batches WHERE id = ?", (body.batch_id,))
    if not batch:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "message": "Batch not found",
            },
        )
    if batch["status"] != "active":
        return JSONResponse(
            status_code=409,
            content={
                "error": "conflict",
                "message": "Can only assign to active batches",
            },
        )
    now = now_utc()
    await db.execute(
        "UPDATE devices SET batch_id = ?, assigned_at = ?, updated_at = ? WHERE id = ?",
        (body.batch_id, now, now, device_id),
    )
    # Backfill unassigned readings from this device after batch start
    await db.execute(
        "UPDATE readings SET batch_id = ? "
        "WHERE device_id = ? AND batch_id IS NULL "
        "AND source_timestamp >= ?",
        (body.batch_id, device_id, batch["started_at"]),
    )
    return await db.query_one("SELECT * FROM devices WHERE id = ?", (device_id,))


@router.post("/{device_id}/unassign", response_model=DeviceResponse)
async def unassign_device(device_id: str, request: Request):
    db = get_db(request)
    device = await db.query_one("SELECT * FROM devices WHERE id = ?", (device_id,))
    if not device:
        return JSONResponse(
            status_code=404,
            content={
                "error": "not_found",
                "message": "Device not found",
            },
        )
    now = now_utc()
    await db.execute(
        "UPDATE devices SET batch_id = NULL, assigned_at = NULL, updated_at = ? WHERE id = ?",
        (now, device_id),
    )
    return await db.query_one("SELECT * FROM devices WHERE id = ?", (device_id,))

File: api/src/routes/devices.ts
import { Hono } from "hono";
import type { AppEnv } from "../app";
import { DeviceCreateSchema, DeviceAssignSchema } from "../models";
import { notFound, conflict, validationError } from "../lib/errors";
import { nowUtc } from "../lib/time";

const devices = new Hono<AppEnv>();

devices.post("/", async (c) => {
  const body = await c.req.json().catch(() => null);
  const parsed = DeviceCreateSchema.safeParse(body);
  if (!parsed.success) return validationError(parsed.error.issues);

  const db = c.env.DB;
  const user = c.get("user");
  const now = nowUtc();
  try {
    await db
      .prepare("INSERT INTO devices (id, name, user_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?)")
      .bind(parsed.data.id, parsed.data.name, user.id, now, now)
      .run();
  } catch (e: any) {
    if (String(e).includes("UNIQUE")) return conflict("Device already registered");
    throw e;
  }
  return c.json(
    await db.prepare("SELECT * FROM devices WHERE id = ?").bind(parsed.data.id).first(),
    201,
  );
});

devices.get("/", async (c) => {
  const user = c.get("user");
  const result = await c.env.DB.prepare("SELECT * FROM devices WHERE user_id = ? ORDER BY created_at DESC").bind(user.id).all();
  return c.json({ items: result.results });
});

devices.post("/claim", async (c) => {
  const db = c.env.DB;
  const user = c.get("user");
  const body = await c.req.json().catch(() => null);
  if (!body?.device_id) return validationError([{ message: "device_id required" }]);

  // Check device exists and is unclaimed
  const device = await db.prepare("SELECT * FROM devices WHERE id = ? AND user_id IS NULL")
    .bind(body.device_id).first();
  if (!device) return notFound("Device not found or already claimed");

  const now = nowUtc();
  await db.batch([
    db.prepare("UPDATE devices SET user_id = ?, updated_at = ? WHERE id = ?")
      .bind(user.id, now, body.device_id),
    db.prepare("UPDATE readings SET user_id = ? WHERE device_id = ? AND user_id IS NULL")
      .bind(user.id, body.device_id),
  ]);

  const updated = await db.prepare("SELECT * FROM devices WHERE id = ?").bind(body.device_id).first();
  return c.json(updated);
});

devices.post("/:deviceId/assign", async (c) => {
  const db = c.env.DB;
  const user = c.get("user");
  const deviceId = c.req.param("deviceId");
  const device = await db.prepare("SELECT * FROM devices WHERE id = ? AND user_id = ?").bind(deviceId, user.id).first();
  if (!device) return notFound("Device");

  const body = await c.req.json().catch(() => null);
  const parsed = DeviceAssignSchema.safeParse(body);
  if (!parsed.success) return validationError(parsed.error.issues);

  const batch = await db.prepare("SELECT * FROM batches WHERE id = ? AND user_id = ?").bind(parsed.data.batch_id, user.id).first<any>();
  if (!batch) return notFound("Batch");
  if (batch.status !== "active") return conflict("Can only assign to active batches");

  const now = nowUtc();
  await db.prepare("UPDATE devices SET batch_id = ?, assigned_at = ?, updated_at = ? WHERE id = ?")
    .bind(parsed.data.batch_id, now, now, deviceId).run();

  // Backfill unassigned readings from this device after batch start
  await db
    .prepare("UPDATE readings SET batch_id = ? WHERE device_id = ? AND batch_id IS NULL AND source_timestamp >= ?")
    .bind(parsed.data.batch_id, deviceId, batch.started_at).run();

  return c.json(await db.prepare("SELECT * FROM devices WHERE id = ?").bind(deviceId).first());
});

devices.post("/:deviceId/unassign", async (c) => {
  const db = c.env.DB;
  const user = c.get("user");
  const deviceId = c.req.param("deviceId");
  const device = await db.prepare("SELECT * FROM devices WHERE id = ? AND user_id = ?").bind(deviceId, user.id).first();
  if (!device) return notFound("Device");

  const now = nowUtc();
  await db.prepare("UPDATE devices SET batch_id = NULL, assigned_at = NULL, updated_at = ? WHERE id = ?")
    .bind(now, deviceId).run();
  return c.json(await db.prepare("SELECT * FROM devices WHERE id = ?").bind(deviceId).first());
});

export default devices;
